### ISSUE TITLE
IBController doesn't consistently select 'paper' trading mode on IB Gateway 959.1b (and other previous versions as well)

### ISSUE THREAD

**Issue by bfreis** [ib-controller/ib-controller#118]
UPDATE-

tl;dr - as explained in the many messages below, **if the file `jts.ini` doesn't exist, IB Controller fails to select Trading Mode = 'paper' in the login screen**; this has been reproduced consistently on my tests.

---

Hello,

For some reason, IB Controller 3.2.0 some times fails to select 'paper' Trading Mode. It's inconsistent - sometimes it does select paper, sometimes it doesn't. When it doesn't, since I'm using my paper credentials, IB Gateway will then show an error message complaining that I used paper user/password but didn't select paper trading mode.

Please see the attached screenshot.

Any ideas?
Thanks!

Here's some additional info.

1) The java process is started like this: `/opt/i4j_jres/1.8.0_60/bin/java -cp /root/Jts/959/jars/*:/opt/IBController/IBController.jar -Xmx256m -Xms256m -Xss1m ibcontroller.IBGatewayController /root/IBController/IBController.ini MY_USERNAME MY_PASSWORD paper` (with redacted user/password).

2) The IB Controller logs seem to show that it correctly understood that I wanted to use my paper account:

```
18:19:13:610 IBController: using default settings provider: ini file is /root/IBController/IBController.ini
18:19:13:614 IBController: using default login manager: getting username and password from args
18:19:13:615 IBController: using default config dialog manager: constructor parameter isGateway=true
18:19:13:615 IBController: using default trading mode manager: constructor parameter args: tradingMode=paper
```

3) However, the following log entry on IB Controller only exists when it works:

```
 IBController: Setting Trading mode = paper
```

I've compared the full logs for one execution that did work, and one that didn't. That line is the single difference in the logs.

4) I've tried different settings on `IBController.ini` for the parameter `TradingMode`. I tried to comment out the line, I tried to set it to 'paper', and I tried to set it to 'live'. It doesn't seem to make any difference - I still observe inconsistent results.

5) The 'launcher.log' file contains the following lines when IB Controller fails to select Trading Mode 'paper', that seem to be related to this issue:

```
2016-10-25 18:57:56.639 [DN] INFO  [JTS-AuthDispatcherS2-16] - NSErrorResponse.PROD_PAPER_INVALID_CHOICE
2016-10-25 18:57:56.642 [DN] INFO  [JTS-AuthDispatcherS2-16] - Connection to server failed: <html>You have selected the Live Trading Mode, but the specified user <br> is a Paper Trading user. <br><br>Please select the correct Trading Mode.</html>
```

---

<img width="706" alt="ibcontroller-issue" src="https://cloud.githubusercontent.com/assets/261623/19710405/e9436710-9ae2-11e6-9326-6cc1f1e2e8a1.png">

**Comment by bfreis** [comment]
OK - right after I posted this I noticed something new.

It seems like it consistently fails the first time I run it. And it seems to consistently work the any subsequent times I run it.

In my setup, I run all this inside a Docker container, that has its filesystem 'reset' every time I kill it. So it's not all that strange that the 'first' execution is different than all the subsequent executions - IB Gateway, or even IB Controller itself, might be writing something to disk on the first execution that changes the behavior for the subsequent executions. That's the single explanation I can imagine - however it still seems like some kind of "bug".

Any ideas?
Thanks

**Comment by bfreis** [comment]
Here's a list of filesystem changes before the 1st execution to after the 1st execution (ie, the 1st failure to select 'paper' Trading Mode):

```
> A /root/.hwid
> C /root/.java
> C /root/.java/fonts
> C /root/.java/fonts/1.8.0_60
> A /root/.java/fonts/1.8.0_60/fcinfo-1-ab5ab1ded859-Linux-4.4.20-moby-en.properties
> C /root/.oracle_jre_usage
> A /root/.oracle_jre_usage/1722dc035deb62ea.timestamp
> C /root/ib-data
> A /root/ib-data/jars
> A /root/ib-data/jars/locales.jar
> A /root/ib-data/jts.ini
> A /root/ib-data/launcher.log
> A /root/ib-data/loginFailFrequency.txt
> C /tmp/hsperfdata_root
> D /tmp/hsperfdata_root/57
> A /tmp/hsperfdata_root/42
```

Here, C = change, A = add, D = delete.

Possibly, the existence/absence of some of these files/directories could be the culprit?

**Comment by bfreis** [comment]
Upon further investigation, it seems like `jts.ini` is the file related to this issue.

For some reason, **if the `jts.ini` file doesn't exist, IB Controller fails to select Trading Mode = 'paper'.** If the file does exist (and it doesn't matter if jts.ini says `tradingMode=l` or `tradingMode=p`), then IB Controller succeeds in selecting Trading Mode = 'paper'.

Any ideas on why or if it's possible to fix the issue?

**Comment by henryshtang** [comment]
I have had the same issue and I get around it by putting in the pre-configured jts.ini in the correct folder. I think you eventually need some initial configuration and so you have to put the jts.ini there anyway.

**Comment by jaredbroad** [comment]
And 960.2a doesn't select paper at all even when configuration file is in place. Works perfectly with live trading. Is there any archive of historical IB installers?

**Comment by rlktradewright** [comment]
@jaredbroad 

I just installed 960.2a and it works perfectly for me in paper trading mode. Can you elaborate on the problem you're encountering please?

@bfreis 

I've just replicated your problem and done some investigation.

The problem arises when the jts.ini files doesn't contain the following lines:

`[IBGateway]`
`ApiOnly=true`

If these lines aren't found, the gateway displays a login form that has a structure that IBController doesn't expect, and it can't find the trading mode selector, which defaults to 'live'. Helpfully, however, it then puts these lines in so that next time you run the gateway, they are there and all is well! Note that IBController doesn't complain about not being able to find the trading mode selector, because only recent versions of TWS/Gateway have it.

Now, when IBController starts it checks for the existence of a jts.ini file, and if it doesn't exist it creates one containing particular information. This is because at the time that IBController 3.2.0 was released, the latest version of TWS (I think it was 956) displayed the login form incorrectly unless jts.ini contained 

`s3store=true`

and this would cause IBController to fail when running TWS because it couldn't find the 'Store settings on server' option. IBController does complain about this because this option has been available for donkeys years, and not finding it is a serious matter.

So the jts.ini file that IBController creates contains what is needed for TWS 956, but doesn't contain what's needed for Gateway 959! In fact, it seems that IB have now fixed the 956 issue, and TWS 960 no longer needs the `s3store=true` line to be present. But they don't seem to have noticed the problem with the Gateway.

It would be a trivial matter to amend IBController to add these extra lines to the skeleton jts.ini file, and maybe I'll get round to it some time, but strictly speaking I'm no longer maintaining IBController so  perhaps someone else would like to do this?

**Comment by jaredbroad** [comment]
@rlktradewright on further investigation it might just be a misreading of the docs: 

_"Note that if the supplied credentials are for a paper trading account, the setting is ignored. For earlier versions of TWS this setting has no effect."_ 

-> I thought this inferred it was effectively an optional parameter. That if I submitted paper credentials when live was selected it would continue as planned on paper - but it instead shows the "Connection to Server" error above @bfreis is trying to get past. Is Trading Mode now a required parameter (either through IBController.ini or the commandline args)?

PS: For posterity QuantConnect's started collecting historical versions of IB Gateway here: If anyone has other versions we can add to the archive please send them to me:
[Standalone Gateway 959.1a](http://data.quantconnect.com/interactive/ibgateway-latest-standalone-linux-x64-v959.1a.sh)
[Standalone Gateway 960.2b ](http://data.quantconnect.com/interactive/ibgateway-latest-standalone-linux-x64-v960.2a.sh)

**Comment by rlktradewright** [comment]
@jaredbroad I noticed that statement last night during my investigations, but the connection with your observation didn't occur to me. Your interpretation is correct, and it was true at the time of writing (for TWS 956 I think, which was the first version that introduced the trading mode), but it appears not to be true for 960 (and perhaps also for some interim versions).

I actually thought it was quite a sensible implementation, but presumably IB have received feedback that it's confusing or something and have 'fixed' it.

**Comment by mateothegreat** [comment]
Invoke like this until patched:

`/bin/bash /opt/IBController/Scripts/IBController.sh 962 -g --mode=paper`

_Log snips.._

```
01:04:58:066 IBController: using default trading mode manager: constructor parameter args: tradingMode=paper
01:04:59:807 IBController: detected frame entitled: IB Gateway; event=Opened
01:04:59:809 IBController: Setting Trading mode = paper
01:04:59:869 IBController: detected frame entitled: IB Gateway; event=Activated
01:04:59:873 IBController: Click button: Login
```

Make sure your **jts.ini** file has `tradingMode=p` under the `[Logon]` section.. viola~!
